### DISCUSSION TITLE
All subsequent tasks getting skipped

### DISCUSSION THREAD

**Discussion by silverbullet1** [apache/airflow#19032] [Q&A] [answered]
So I have this section of my DAG,
![image](https://user-images.githubusercontent.com/15612277/137618063-6f5bf6f0-2dee-4613-b188-c77b687b82ae.png)
The problem is that all my tasks are getting skipped even though I have mapped the connections correctly in my PythonBranchOperator. I am basically expecting create_clickhouse_tables to execute and only the other branch to skip.
My code is somewhat like this:

> if_table_exists >> [create_clickhouse_tables, compare_schema]
    compare_schema >> [check_backward_compatibility, create_clickhouse_tables]
    check_backward_compatibility >> [create_clickhouse_tables, check_rebuild_flag]
    check_rebuild_flag >> [create_clickhouse_tables, err]
    create_clickhouse_tables >> complete

In the python callable of if_tables_exist, I am returning "create_clickhouse_tables" which is the task id of the task that I want to execute(you can assume that it is hardcoded for now).

Not sure why this is happening.

**Comment by josh-fell** [comment]
Tough to say without seeing a little more the DAG code itself rather than just the task dependencies. Would you be able to provide some more code context? Scrubbing what you feel is necessary is completely fine.

**Reply by josh-fell** [reply]
Is `if_table_exists` also a branching operator? If so, it may be the culprit.

**Reply by silverbullet1** [reply]
Hi @josh-fell, Yes it is a branching operator because that's my requirement. If they don't exist then I want to do create_clickhouse_tables, otherwise I want to check other things. Why is this the culprit? and how can I fix it ?

**Reply by josh-fell** [reply]
Looking at the screenshot, _all_ of the downstream tasks from the `if_table_exists` task are skipped.  My suspicion is that this task is not returning immediate, downstream task IDs to follow ("compare_schema" or "create_clickhouse_tables" in this case).  Hardcoding "create_clickhouse_tables" as the output for the `compare_schema` task as you mention won't matter if `compare_schema` is itself skipped.

What's the callable logic for the `if_table_exists` task?

**Reply by silverbullet1** [reply]
Hi @josh-fell,

callable logic for `if_table_exists` : 
```
def check_ch(ti):
    if ti.xcom_pull( some_business_logic ):
        return "create_clickhouse_tables"
    else:
        return "compare_schema"
```

These return values are the task ID's which I want to execute. I am expecting create_clickhouse_tables to be called after "if_table_exists" because my python callable indeed returns that task id.

**Reply by silverbullet1** [reply]
Also, I meant i am hardcoding "create_clickhouse_tables" as the output for "if_tables_exist" and not "compare_schema", sorry about that.

**Reply by josh-fell** [reply]
Hmm. And no special `trigger_rule` set on "compare_schema" or "create_clickhouse_tables" tasks? What Airflow version are you using?

Are you able to broadly show more DAG code? Trying to get a sense of the larger picture.

**Comment by silverbullet1** [comment]
I am using airflow v2.1.3.
No special trigger_rule set.
Here is some more relevant code :


```
# This is just before if_table_exists task. This operator's result is pushed into xcom, which is what I check in the if_table_exists task, 
# and it evaluates correctly, but does not trigger anything further.
    check_ch = ClickHouseOperator(
            task_id='check_ch',
            database='default',
            sql="...",
            # Result is pushed to xcom
            clickhouse_conn_id='clickhouse',
        )

    if_table_exists = BranchPythonOperator(
            task_id='if_table_exists',
            provide_context=True,
            python_callable=if_table_exists_callable
        )

    compare_schema = BranchPythonOperator(task_id = "compare_schema", python_callable = compare_schema_callable)

    check_backward_compatibility = BranchPythonOperator(task_id = "check_backward_compatibility", python_callable = check_backward_compatiblity_callable)

    check_rebuild_flag = BranchPythonOperator(task_id = "check_rebuild_flag", python_callable = check_rebuild_flag_callable)

     create_clickhouse_tables = ClickHouseOperator(
            task_id='create_clickhouse_tables',
            database='default',
            sql="{{ .. }",
            clickhouse_conn_id='clickhouse'
        )

    err = DummyOperator(task_id='job_failed')

    complete = DummyOperator(task_id='All_jobs_completed')


```

Here are the callables(most of them are empty for now)

```

def if_table_exists_callable(ti):
    if ti.xcom_pull(somekey = somevalue)
        return "create_clickhouse_tables"
    else:
        return "compare_schema"

def check_backward_compatiblity_callable():
    pass
    # return "is_rebuild_flag_set"

def check_rebuild_flag_callable():
    pass
    # return "job_failed"

def compare_schema_callable():
    pass
    # return "is_backward_compatible"

```
The sequence is already mentioned in the OP.

I tried hardcodng the return values but it did not work. I am basically trying to implement one scenario first so I have not implemented these functions yet because my first function(if_table_exists_callable) itself isn't working.

**Comment by josh-fell** [comment, marked as answer]
The "create_clickhouse_tables" task needs to have a `trigger_rule` of `TriggerRule.NONE_FAILED` (aka "none_failed") for it to be executed when appropriate. By default, tasks have a `trigger_rule` value of `TriggerRule.ALL_SUCCESS` (aka "all_success"). In this case each of upstream tasks to "create_clickhouse_tables" could have a "skipped" or "success" state depending on branching logic. Setting the `trigger_rule` value to `TriggerRule.NONE_FAILED` will allow the task to execute assuming no upstream tasks have failed.

```python
from datetime import datetime

from airflow.decorators import dag, task
from airflow.operators.dummy import DummyOperator
from airflow.operators.python import BranchPythonOperator
from airflow.utils.trigger_rule import TriggerRule


@task
def check():
    return "some_output"


def if_table_exists_callable(ti):
    print(ti.xcom_pull(task_ids="check"))
    if ti.xcom_pull(task_ids="check") == "some_output":
        return "create_clickhouse_tables"

    return "compare_schema"


def check_backward_compatiblity_callable():
    return "check_rebuild_flag"


def check_rebuild_flag_callable():
    return "job_failed"


def compare_schema_callable():
    return "check_backward_compatibility"


@dag(
    dag_id="branch_test",
    start_date=datetime(2021, 1, 1),
    schedule_interval=None,
)
def branch_test():
    check_ch = check()

    if_table_exists = BranchPythonOperator(
        task_id="if_table_exists", python_callable=if_table_exists_callable
    )

    compare_schema = BranchPythonOperator(task_id="compare_schema", python_callable=compare_schema_callable)

    check_backward_compatibility = BranchPythonOperator(
        task_id="check_backward_compatibility", python_callable=check_backward_compatiblity_callable
    )

    check_rebuild_flag = BranchPythonOperator(
        task_id="check_rebuild_flag", python_callable=check_rebuild_flag_callable
    )

    create_clickhouse_tables = DummyOperator(
        task_id="create_clickhouse_tables", trigger_rule=TriggerRule.NONE_FAILED
    )

    err = DummyOperator(task_id="job_failed")

    complete = DummyOperator(task_id="All_jobs_completed")

    check_ch >> if_table_exists >> [compare_schema, create_clickhouse_tables]
    compare_schema >> [check_backward_compatibility, create_clickhouse_tables]
    check_backward_compatibility >> [check_rebuild_flag, create_clickhouse_tables]
    check_rebuild_flag >> [err, create_clickhouse_tables]
    create_clickhouse_tables >> complete


dag = branch_test()
```
![image](https://user-images.githubusercontent.com/48934154/138382893-79ac77d6-51e3-4243-8515-f64bbac32f42.png)

**Reply by silverbullet1** [reply]
Awesome! Thanks. I did not know about trigger rule set. This works :)
